### PR TITLE
Eliminates subscription-manager exec on every Puppet run

### LINKED ISSUE #90: Read facts from /etc/yum.repos.d/redhat.repo instead of subscription-manager
As stated by @alexjfisher in #53 `subscription-manager` might lead to long puppet runs and - in case RHSM/Satellite 6 is down - might lead to issues when relying on the resulting facts:

If the Satellite server (or RHSM) is not accessible or sends http errorcodes.

1. Not accessible: Puppet runs are stale until `subscription-manager` times out (default timeout is 180s per http request and `subscription-manager repos --list-enabled` triggers ~7 requests)
2. Satellite/RHSM frontend answers with http errorcode: fact returns empty array -> if someone (like we do) evaluates this fact to enable repos, Puppet tries to enable all repos (again - not a big issue but def. not a nice behaviour). 

We were thinking about a caching mechanism which addresses these two issues.

Instead of calling `subscription-manager` or `yum` the file `/etc/yum.repos.d/redhat.repo` should be evaluated

 ".. or search for `enabled = 1` and all sections without `enabled`

This should also address both issues.

## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: >=4
- Ruby: generic
- Distribution: RHEL
- Module version: 3.2.1-rc0/master

## How to reproduce (e.g Puppet code you use)

As described in README.md

## What are you seeing

Slow puppet runs (>10s for `subscription-manager`), issues when Satellite 6 is not available or - even worse - partially available.

### LINKED ISSUE #92: Don't enforce latest package to be installed as default value
The `ensure` default value for the `subscription-manager` is `latest`. This causes slowed down puppet runs (caused by `yum check-update`).

https://github.com/voxpupuli/puppet-rhsm/blob/master/manifests/init.pp#L75

I suggest to change the default to `installed`.

### PR DESCRIPTION
**Pull request by speer**
#### Pull Request (PR) description
The idea behind this PR is to get rid of `subscription-manager` executions on every Puppet-run. If there are issues with subscription-manager, (ex. Satellite down or similar) these executions block all the Puppet runs on all connected systems. In those cases, the whole Puppet infrastructure becomes unusable, since during Puppet runs, all calls of "subscription-manager identity, subscription-manager list, subscription-manager repos" hang and lead to almost never finishing Puppet agent runs. There are basically no timeouts of some of these calls (neither server_timeout in rhsm.conf leads to improvements of the situation).
Also we found, that in such situations, the module had tried to run "auto attach" on all of our systems. This is probably because the onlyif condition of RHSM-subscribe triggered somehow...

Because of the reasons explained above, we found, this module should not rely on subscription-manager anymore for verifying the systems current state and use it only for registering systems, assigning/removing subscriptions and enabling/disabling repos.

With this PR, the module relies on entitlement certificates and the redhat.repo file, which are kept up to date periodically by rhsmcertd. Of course I agree, that there might occur race conditions, where these sources of truth could not always be as reliable as subscription-manager executions, but I find the trade-off justifiable, because of the issues explained above.

Partly, this problem has already been addressed by @kallies in #91 but this was limited to the executions of the rhsm facts only. So the current PR supersedes #91 

In this PR I further removed the `rhsm::repo` type, which had already been replaced by `rh_repo`. I see no need of supporting two types that do the same, just differently.
Also I removed the `pool` parameter in init.pp, since I found that one should use `rh_subscription` for that.
Because of already touching the parameter list of init.pp, I also added the types contributed by @kallies in #86 

Note, that this PR contains changes that break backward-compatibility.

#### This Pull Request (PR) fixes the following issues
Fixes #90
Fixes #92

### CODE REVIEW

File: lib/puppet/provider/rh_subscription/redhat.rb [hunk 1]
```diff
@@ -1,41 +1,40 @@
 # @author Craig Dunn crayfishx/puppet-rhsm
+# @author Stefan Peer speer/puppet-rhsm
+#   (do not run subscription-manager on every Puppet run)
 Puppet::Type.type(:rh_subscription).provide(:redhat) do
   commands rhsm: '/usr/sbin/subscription-manager'
+  commands rct: '/usr/bin/rct'
```
Comment (vinzent) [inline lib/puppet/provider/rh_subscription/redhat.rb]: for those that never heard of this command:

> rct - Displays information (headers) about or size and statistics of a entitlement, product, or identity certificate used by Red Hat Subscription Manager.

File: lib/puppet/provider/rh_subscription/redhat.rb [hunk 2]
```diff
@@ -1,41 +1,40 @@
...
+
   mk_resource_methods
 
   def self.subscriptions
     subs = {}
-    rhsm('list', '--consumed').split("\n\n").each do |blk|
```
Comment (vinzent) [inline lib/puppet/provider/rh_subscription/redhat.rb]: could you document the method? especially some return value example

File: lib/puppet/provider/rh_subscription/redhat.rb [hunk 3]
```diff
@@ -1,41 +1,40 @@
...
+      next if cert =~ %r{-key.pem$}
+      cert_raw_data = rct('cat-cert', cert)
       sub_data = {}
-      blk.split("\n").each do |line|
-        element = line.match(%r{^([^:]+):[^\w]+(\w.*)})
-        next unless element.is_a?(MatchData)
-        sub_data[element[1]] = element[2]
+      section = nil
+      cert_raw_data.split("\n").each do |line|
+        if line =~ %r{^([^:\s]+):$}
```
Comment (vinzent) [inline lib/puppet/provider/rh_subscription/redhat.rb]: what are you doing here?

Comment (vinzent) [reply lib/puppet/provider/rh_subscription/redhat.rb]: maybe some example of the output of rct would help.

File: manifests/init.pp [hunk 4]
```diff
@@ -54,27 +51,26 @@
...
-  $serverport                        = 443,
-  $ca_cert_dir                       = '/etc/rhsm/ca/',
-  $repo_ca_cert_filename             = 'redhat-uep.pem',
-  $repo_ca_cert_source               = undef,
-  $manage_repos                      = 1,
-  $full_refresh_on_yum               = 0,
-  $package_ensure                    = 'latest',
-  Array[String[1]] $enabled_repo_ids = [],
-) {
+  Optional[String[1]]    $rh_user               = undef,
```
Comment (vinzent) [inline manifests/init.pp]: Just FYI: back in the days we explicitly didn't use data types potentially stay compatible with the old puppet provided by Redhat with Satellite 6.

File: lib/puppet/provider/rh_subscription/redhat.rb (hunk 3 above)
Comment (speer) [reply lib/puppet/provider/rh_subscription/redhat.rb]: Thank's for your feedback! Added some documentation, hope that explains it better :)

File: manifests/init.pp [hunk 5]
```diff
@@ -54,27 +51,26 @@
...
+  Optional[String[1]]    $rh_password           = undef,
```
Comment (bastelfreak) [inline manifests/init.pp]: stricter datatypes \o/

File: lib/puppet/provider/rh_repo/redhat.rb [hunk 6]
```diff
@@ -1,18 +1,19 @@
...
     repos = {}
-    rhsm('repos', '--list').split("\n\n").each do |blk|
-      repo_data = {}
-      blk.split("\n").each do |line|
-        element = line.split(%r{:})
-        repo_data[element.shift] = element.join.lstrip
-      end
-      repos[repo_data['Repo ID']] = {
-        enabled: repo_data['Enabled']
+    repo_file = Puppet::Util::IniConfig::PhysicalFile.new('/etc/yum.repos.d/redhat.repo')
```
Comment (kallies) [inline lib/puppet/provider/rh_repo/redhat.rb]: This will improve reliability and performce!

File: lib/puppet/provider/rh_subscription/redhat.rb [hunk 7]
```diff
@@ -1,41 +1,121 @@
...
 
+  # Retrieves all current subscriptions, by inspecting the system's entitlement
+  # certificates under /etc/pki/entitlement, via the "rct" tool.
+  #
+  # rct - Displays information (headers) about or size and statistics of a entitlement,
+  # product, or identity certificate used by Red Hat Subscription Manager.
+  #
+  # Example output of rct cat-cert (not complete):
+  #
+  # $ rct-cat cert /etc/pki/entitlement/123421432423233332.pem
```
Comment (kallies) [inline lib/puppet/provider/rh_subscription/redhat.rb]: It's `rct cat-cert`

Comment (speer) [reply lib/puppet/provider/rh_subscription/redhat.rb]: corrected typo

File: manifests/init.pp (hunk 4 above)
Comment (kallies) [reply manifests/init.pp]: As the type/provider code does not work with Puppet 3 Agents, this is fine for Satellite >= 6.4 (cornercase: Puppet 4/5 Server and Puppet 3 Agent)

**Comment by kallies** [comment]
@bastelfreak @ekohl does this PR still need feedback?

**Comment by ekohl** [comment]
I hadn't looked at it before. For the future, splitting the data types and other cleanups into a separate commit makes reviewing easier. You don't have to split it now. I must admit I don't know enough about subscription manager to properly review this though :(

File: lib/puppet/provider/rh_subscription/redhat.rb [hunk 8]
```diff
@@ -1,41 +1,121 @@
...
+        next if section.nil?
+        sub_data[section][Regexp.last_match(1).strip] = Regexp.last_match(2).strip if line =~ %r{^\s+([^:]+):(.+)$}
       end
-      subs[sub_data['Pool ID']] = {
-        name: sub_data['Subscription Name'],
-        active: sub_data['Active'],
-        serial: sub_data['Serial']
+      subs[sub_data['Certificate']['Pool ID']] = {
+        name: sub_data['Order']['Name'],
+        serial: sub_data['Certificate']['Serial']
```
Comment (kallies) [inline lib/puppet/provider/rh_subscription/redhat.rb]: Rubocop mentions the missing comma at the end of the line

File: lib/puppet/provider/rh_subscription/redhat.rb [hunk 9]
```diff
@@ -1,41 +1,121 @@
...
-    end
-  end
-
   def self.instances
     subscriptions.map do |pool, data|
       new(
         ensure: :present,
         name:   pool,
-        active: active?(data[:active]),
         serial: data[:serial]
```
Comment (kallies) [inline lib/puppet/provider/rh_subscription/redhat.rb]: Rubocop mentions the missing comma at the end of the line

Comment (speer) [reply lib/puppet/provider/rh_subscription/redhat.rb]: If I add the comma, TravisCI build fails...

Offenses:

lib/puppet/provider/rh_subscription/redhat.rb:108:50: C: Style/TrailingCommaInLiteral: Avoid comma after the last item of a hash. (https://github.com/bbatsov/ruby-style-guide#no-trailing-array-commas)

        serial: sub_data['Certificate']['Serial'],

                                                 ^

lib/puppet/provider/rh_subscription/redhat.rb:119:30: C: Style/TrailingCommaInArguments: Avoid comma after the last parameter of a method call. (https://github.com/bbatsov/ruby-style-guide#no-trailing-params-comma)

        serial: data[:serial],

File: lib/puppet/provider/rh_subscription/redhat.rb (hunk 8 above)
Comment (speer) [reply lib/puppet/provider/rh_subscription/redhat.rb]: If I add the comma, TravisCI build fails...

Offenses:

lib/puppet/provider/rh_subscription/redhat.rb:108:50: C: Style/TrailingCommaInLiteral: Avoid comma after the last item of a hash. (https://github.com/bbatsov/ruby-style-guide#no-trailing-array-commas)

        serial: sub_data['Certificate']['Serial'],

                                                 ^

lib/puppet/provider/rh_subscription/redhat.rb:119:30: C: Style/TrailingCommaInArguments: Avoid comma after the last parameter of a method call. (https://github.com/bbatsov/ruby-style-guide#no-trailing-params-comma)

        serial: data[:serial],

File: lib/puppet/provider/rh_subscription/redhat.rb (hunk 9 above)
Comment (ekohl) [reply lib/puppet/provider/rh_subscription/redhat.rb]: Yes, in Rubocop we have different rules. Personally I'd prefer trailing commas here as well but let's not change rubocop rules as part of this PR.

**Pull request merged**